File: BookLibrary/Server/ILibraryService.fs
namespace BookLibrary.Server

open System.Runtime.Serialization
open System.ServiceModel


[<ServiceContract(SessionMode = SessionMode.Required)>]
type ILibraryService =
    [<OperationContract(IsInitiating = true, IsTerminating = false)>]
    abstract EnterLibrary: userId:int -> userName:string -> unit

    [<OperationContract(IsInitiating = false)>]
    abstract AddBook: book:Book -> unit

    [<OperationContract(IsInitiating = false)>]
    abstract GetBook: bookId:int -> Book

    [<OperationContract(IsInitiating = false)>]
    abstract GetBooks: bookAuthor:string -> seq<Book>

    [<OperationContract(IsInitiating = false)>]
    abstract TakeBook: bookId:int -> unit

    [<OperationContract(IsInitiating = false)>]
    abstract ReturnBook: bookId:int -> unit

    [<OperationContract(IsInitiating = false, IsTerminating = true)>]
    abstract SaveChanges: unit -> unit

    [<OperationContract(IsTerminating = true)>]
    abstract Leave: unit -> unit

File: BookLibrary/Server/ILibraryService.cs
using System.Collections.Generic;
using System.ServiceModel;

namespace Server
{
    [ServiceContract]
    public interface ISaveChangesOperationCallback
    {
        [OperationContract(IsOneWay = true)]
        void OnSaveChanges();
    }

    [ServiceContract(CallbackContract = typeof(ISaveChangesOperationCallback), SessionMode = SessionMode.Required)]
    public interface ILibraryService
    {
        [OperationContract(IsInitiating = true, IsTerminating = false)]
        void EnterLibrary(int userId, string userName);

        [OperationContract(IsInitiating = false)]
        void AddBook(Book book);

        [OperationContract(IsInitiating = false)]
        Book GetBook(int bookId);

        [OperationContract(IsInitiating = false)]
        IEnumerable<Book> GetBooks(string authorName);

        [OperationContract(IsInitiating = false)]
        void TakeBook(int bookId);

        [OperationContract(IsInitiating = false)]
        void ReturnBook(int bookId);

        [OperationContract(IsInitiating = false, IsOneWay = true)]
        void SaveChanges();

        [OperationContract(IsInitiating = false, IsTerminating = true, IsOneWay = true)]
        void Leave();
    }
}